papas/pfalgo/pfblockbuilder.py

```diff
@@ -24,44 +24,39 @@ class PFBlockBuilder(BlockBuilder):
-            builder = PFBlockBuilder(pfevent, ruler)
+            builder = PFBlockBuilder(papasevent, ruler)
             for b in builder.blocks.itervalues() :
                 print b
     '''
-    def __init__(self,  pfevent, ruler, history_nodes = None):
+    def __init__(self,  uniqueids, papasevent, ruler, subtype ='r'):
         '''
-        pfevent is event structure inside which we find
+       
             tracks is a dictionary : {id1:track1, id2:track2, ...}
-            ecal is a dictionary : {id1:ecal1, id2:ecal2, ...}
-            hcal is a dictionary : {id1:hcal1, id2:hcal2, ...}
+            ecals is a dictionary : {id1:ecal1, id2:ecal2, ...}
+            hcals is a dictionary : {id1:hcal1, id2:hcal2, ...}
             get_object() which allows a cluster or track to be found from its id
+            history is a dictionary of Nodes : { id:Node1, id: Node2 etc}
+                containing the simulation history 
+                A Node contains the id of an item (cluster, track, particle etc)
+                and says what it is linked to (its parents and children)
+                if hist_nodes is provided it will be added to with the new block information
+                If hist_nodes is not provided one will be created, it will contain nodes
+                corresponding to each of the tracks, ecal etc and also for the blocks that
+                are created by the event block builder.
         ruler is something that measures distance between two objects eg track and hcal
             (see Distance class for example)
             it should take the two objects as arguments and return a tuple
             of the form
                 link_type = 'ecal_ecal', 'ecal_track' etc
                 is_link = true/false
                 distance = float
-        history_nodes is an optional dictionary of Nodes : { id:Node1, id: Node2 etc}
-            it could for example contain the simulation history nodes
-            A Node contains the id of an item (cluster, track, particle etc)
-            and says what it is linked to (its parents and children)
-            if hist_nodes is provided it will be added to with the new block information
-            If hist_nodes is not provided one will be created, it will contain nodes
-            corresponding to each of the tracks, ecal etc and also for the blocks that
-            are created by the event block builder.
         '''
         
-        #given a unique id this can return the underying object
-        self.pfevent = pfevent
-
-        # collate all the ids of tracks and clusters and, if needed, make history nodes
-        uniqueids = []
-        uniqueids = list(pfevent.tracks.keys()) + list(pfevent.ecal_clusters.keys()) + list(pfevent.hcal_clusters.keys()) 
+        
         uniqueids = sorted(uniqueids)
-
-        self.history_nodes = history_nodes
-        if history_nodes is None:
+        self.papasevent = papasevent
+        self.history_nodes = papasevent.history
```

Comment:
is there a real need for an additional attribute? could use papasevent.history directly